### PR TITLE
Refactor and Enable batching for both simple and openai

### PR DESCRIPTION
**Pull request by xwu99**
Completed:
- An enhanced version to replace #149, support both simple and openai protocols.
- Refactor PredictorDeployment: `__call__` for simple and `openai_call` for openai
- Support Continuous Batching (use vllm), Static Batching (single request multiple prompts) and Dynamic Batching (ray serve batching)

TODO:
- Support static batching output for openai
- Docs

### CODE REVIEW

File: llm_on_ray/inference/deepspeed_predictor.py [hunk 1]
```diff
@@ -244,20 +244,34 @@ def streaming_generate(self, prompt, streamer, **config):
         for worker in self.prediction_workers[1:]:
             worker.streaming_generate.remote(inputs_ref, self._create_dummy_streamer(), **config)
 
-    def generate(self, prompt, **config):
-        input_ids, input_length = self.tokenize_inputs(prompt)
+    def generate(
```
Comment (carsonwang) [inline llm_on_ray/inference/deepspeed_predictor.py:247]: Can we simplify this method by always accepting prompts as List[str] like below? We can write another method that accepts a prompt: str and returns `self.generate([prompt], **config)[0]`

```suggestion
    def generate(
        self, prompts: List[str], **config
    ) -> List[GenerateResult]:
        input_ids, input_length = self.tokenize_inputs(prompts)
        inputs_ref = ray.put(input_ids)
        gen_tokens = ray.get(
            [worker.generate.remote(inputs_ref, **config) for worker in self.prediction_workers]
        )[0]

        decode_result = self.tokenizer.batch_decode(gen_tokens, skip_special_tokens=True)

        return [
            GenerateResult(
                text=decode_result[i],
                input_length=input_length,
                generate_length=gen_tokens.size()[1] - input_length,
            )
            for i in range(len(prompts))
        ]
```

File: llm_on_ray/inference/transformer_predictor.py [hunk 2]
```diff
@@ -115,20 +116,34 @@ def streaming_generate(self, prompt, streamer, **config):
             **config,
         )
 
-    def generate(self, prompt, **config):
+    def generate(
```
Comment (carsonwang) [inline llm_on_ray/inference/transformer_predictor.py:119]: same here.

File: llm_on_ray/inference/predictor_deployment.py [hunk 3]
```diff
@@ -93,145 +104,253 @@ async def consume_streamer_async(self, streamer: TextIteratorStreamer):
...
-                return JSONResponse(status_code=400, content="Invalid prompt format.")
-        else:
-            prompts.append(text)
-
-        if not streaming_response:
-            if self.use_vllm:
-                return await self.predictor.generate_async(prompts, **config)
+    # Handle streaming, only support single prompt
+    async def handle_streaming(self, prompt: str, config: Dict[str, Any]):
+        if isinstance(prompt, list):
```
Comment (carsonwang) [inline llm_on_ray/inference/predictor_deployment.py:109]: Update `prompt: str` to include list in the method definition.

File: llm_on_ray/inference/predictor_deployment.py [hunk 4]
```diff
@@ -93,145 +104,253 @@ async def consume_streamer_async(self, streamer: TextIteratorStreamer):
...
+                    )
+                )
         if self.use_deepspeed:
-            self.predictor.streaming_generate(prompts, self.streamer, **config)
-            return StreamingResponse(
-                self.consume_streamer(), status_code=200, media_type="text/plain"
-            )
+            self.predictor.streaming_generate(prompt, self.streamer, **config)
+            if not self.use_openai:
+                yield StreamingResponse(
```
Comment (carsonwang) [inline llm_on_ray/inference/predictor_deployment.py:130]: I feel this file is too complex, too many if else. Can we further simplify this? For example, if we are returning ModelResponse for openai, can we return SimpleModelResponse for simple serving server, but the logic to create the response can be the same. So you can create a method to generate the response without if else every where? Also, can we refactor for the predictor classes so we don't need to check which predictor we are using? For example, call the same interface for streaming.

Comment (xwu99) [reply llm_on_ray/inference/predictor_deployment.py:130]: The primary purpose of the PR is to add dynamic batching using the original code structure. If have concerns here, I would suggest open another PR to do the refactoring and leave this PR to add dynamic batching later.

File: llm_on_ray/inference/deepspeed_predictor.py (hunk 1 above)
Comment (xwu99) [reply llm_on_ray/inference/deepspeed_predictor.py:247]: thanks, I will add to the refactor plan to always use list as input.

File: llm_on_ray/inference/predictor_deployment.py [hunk 5]
```diff
@@ -93,145 +104,253 @@ async def consume_streamer_async(self, streamer: TextIteratorStreamer):
...
+        logger.info(f"Handling dynamic batch (size={len(requests)}) ...")
+
+        # batch prompts that configs are the same and also save request index
+        batched_prompts: Dict[str, Tuple[Union[str, List[str]], List[int]]] = {}
+        for i, request in enumerate(requests):
+            prompt = request[0]
+            config = request[1]
+            # sort the config by key and convert to str to ensure the key is unique
+            key = str(dict(sorted(config.items())))
+            batched_prompts[key] = batched_prompts.get(str(config), ([], []))
```
Comment (KepingYan) [inline llm_on_ray/inference/predictor_deployment.py:209]: This should be str(dict(sorted(config.items()))), now if config is different from sorted(config.items()), batched_prompts[key] will have only one prompt.

File: README.md [hunk 6]
```diff
@@ -116,6 +116,7 @@ The following are detailed guidelines for pretraining, finetuning and serving LL
 * [Deploy and Serve LLMs on Intel CPU/GPU/Gaudi](docs/serve.md)
 * [Deploy and Serve LLMs with Deepspeed](docs/serve_deepspeed.md)
 * [Deploy and Serve LLMs with BigDL-LLM](docs/serve_bigdl.md)
+* [Enable Continuous, Static and Dynamic Batching](docs/serve_batching.md)
```
Comment (carsonwang) [inline README.md:119]: Let's remove this and the file for now.

**Comment by xwu99** [comment]
- Add tests
- Add max_batch_size parameter
- Add agent tools from main branch

Pls review. thanks!

**Review by KepingYan — APPROVED** [review]
LGTM

**Pull request merged**